File: algorithms/0141.linked_list_cycle/solution.ts
interface ListNode {
  val: number
  next: ListNode | null
}

// 141. Linked List Cycle (Easy)
// https://leetcode.com/problems/linked-list-cycle

// Double pointer solution
const hasCycle = (head: ListNode | null): boolean => {
  if (!head) return false

  let slowHead: ListNode | null | undefined = head
  let fastHead: ListNode | null | undefined = head

  while (slowHead && fastHead) {
    slowHead = slowHead.next
    fastHead = fastHead.next?.next

    if (slowHead === fastHead) return true
  }

  return false
}

File: algorithms/0141.linked_list_cycle/solution.go
package problems

type ListNode struct {
	Val  int
	Next *ListNode
}

// 141. Linked List Cycle (Easy)
// https://leetcode.com/problems/linked-list-cycle

// Double pointer solution
func HasCycle(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return false
	}

	slowHead := head
	fastHead := head

	for fastHead != nil && fastHead.Next != nil {
		slowHead = slowHead.Next
		fastHead = fastHead.Next.Next

		if fastHead == slowHead {
			return true
		}
	}

	return false
}
